pkg/apis/storage/types.go

```diff
@@ -151,7 +151,9 @@ type VolumeAttachmentSource struct {
 	// +optional
 	PersistentVolumeName *string
 
-	// Placeholder for *VolumeSource to accommodate inline volumes in pods.
+	// Copy of VolumeSource to accommodate inline volumes in pods.
+	// +optional
+	InlineVolumeSource *api.CSIPersistentVolumeSource
```

Comment:
It would be useful to include `VolumeSource` changes too (`Pod.Spec.Volumes`), because I doubt that it will contain `CSIPersistentVolumeSource`. It's likely to be `CSIVolumeSource` there.

Reply:
I can do that, The CSIVolumeSource will be the exact same object as `CSIPersistentVolumeSource` though, this would be strictly a naming change. We will also need to explicitly not add it to the `VolumeSource` object though because we do not want to support inline volumes natively with CSI.

Reply:
I looked into it and it seems like the convention is not super clear, for example `GCEPersistentDiskVolumeSource` is the object used in both `VolumeSource` and `PersistentVolumeSource`. I don't think it is worth creating a carbon copy of the `CSIPersistentVolumeSource` and naming it `CSIVolumeSource` just to put it in this one field.

I believe it's fine to leave it as a PersistentVolumeSource object.

Let me know if you disagree, the change is not difficult but I want to try minimize confusion to future maintainers and I think this is the best way. Maybe I should enhance the comment above `InlineVolumeSource` to specify that this is not a PersistentVolume but we are just using this type and it is a misnomer in this case?

Reply:
I agree that the consistency could be better. However, "Persistent" in `GCEPersistentDiskVolumeSource` comes from GCE's Persistent Disk, not from Kubernetes PersistentVolume(Source).

The difference between volume sources in PVs and Pods (such as `PV.*.ISCSIPersistentVolumeSource` and `Pod.*.ISCSIVolumeSource`) are usually secrets. `ISCSIPersistentVolumeSource` can refer to a secret in any namespace (`SecretReference`), while `ISCSIVolumeSource` in a Pod can refer only to secrets in the same namespace (`LocalObjectReference`).

@liggit is usually opinionated in these cases. IMO, CSI should follow the same pattern as the other volume plugins and not allows users to reference any secret in the system in their pods.

Reply:
I agree that the consistency could be better. However, "Persistent" in `GCEPersistentDiskVolumeSource` comes from GCE's Persistent Disk, not from Kubernetes PersistentVolume(Source).

The difference between volume sources in PVs and Pods (such as `PV.*.ISCSIPersistentVolumeSource` and `Pod.*.ISCSIVolumeSource`) are usually secrets. `ISCSIPersistentVolumeSource` can refer to a secret in any namespace (`SecretReference`), while `ISCSIVolumeSource` in a Pod can refer only to secrets in the same namespace (`LocalObjectReference`).

@liggitt  is usually opinionated in these cases. IMO, CSI should follow the same pattern as the other volume plugins and not allows users to reference any secret in the system in their pods.

Reply:
@liggitt's input here would be valuable. For context we are trying to serve an in-tree plugin referenced inline in a pod (e.g. a GCE PD, AWS EBS, etc. volume referenced inline in a pod) using an out-of-tree driver (e.g. GCE PD CSI driver, AWS EBS CSI driver, etc.). The CSI interface purposefully only has a `PersistentVolumeSource` no `VolumeSource` so CSI drivers can't be referenced directly inline. But since the existing in-tree volumes can, we need to find a way to map them for the migration.

Reply:
a few questions to get my bearings:
* what uses the PersistentVolumeName field on VolumeAttachment objects today to traverse to the PV object to look up the PersistentVolumeSource?
* what component do you anticipate being responsible for looking at a Pod volume source and knowing how to transform it into a corresponding CSIPersistentVolumeSource?

the VolumeAttachment object currently works by referencing the name of the PV object containing the canonical PersistentVolumeSource. Is there a reason we wouldn't follow the same approach for pod inline volumes and reference the pod name/namespace (and maybe uid) and the volume name? is there a benefit to duplicating the information here and risking it being stale or becoming detached from the driving pod?

Reply:
`what uses the PersistentVolumeName field on VolumeAttachment objects today to traverse to the PV object to look up the PersistentVolumeSource?`
* The external attacher will GET the PV and grab the csiSource from it to get information it needs to call the CSI attach

`what component do you anticipate being responsible for looking at a Pod volume source and knowing how to transform it into a corresponding CSIPersistentVolumeSource?`
The Volume Plugin responsible for that Pod volume source, see [WIP] design https://github.com/kubernetes/community/pull/2199/files#diff-a5b9eff044138c5af66d85efc24c7cacR203

The problem with referencing pod information in the VolumeAttachment object is that then the "knowledge" of how to translate from arbitrary [read: eventually all] pod volume sources to a CSIPersistentVolumeSource would need to be somehow baked into the CSI external-attacher which I think we would rather avoid.